### DISCUSSION TITLE
RNode transmission behavior (collision detection?)

### DISCUSSION THREAD

**Discussion by SebastianObi** [markqvist/Reticulum#327] [Help & Questions]
Hi,

I have a question which I have not found an answer in the docu or other discussions. 

How exactly does the current transmission behavior of reticulum work with an rnode? I have the problem that if I use more than 2 rnodes it gets significant problems with the data transmission. Which is almost or no longer possible.

Does the RNode use CSMA/CA or another method?

When i tried to use a node as a transport node between tcp and lora. Then I connected a lora client to it, messages, sync and also pages could be transferred. As soon as I used a 2nd nomadnet client with rnode. Which wanted to view pages at the same time, no data transfer worked anymore.

Maybe someone can say something about this behavior. Or I am doing something wrong.

Thanks

**Comment by markqvist** [comment]
The RNode firmware uses CSMA by default. It should not transmit when something else is active on the channel. If on a really slow channel, can it be you are simply exceeding the channel capacity of whatever configured parameters you are using?

**Reply by SebastianObi** [reply]
I made some tests:

Scenario:
3x Raspberry Pi with current debian and current rns+nomadnet.
3x RNode with current firmware
Node 1 host some pages.
Node 2+3 are clients which want to view the pages.

Interface settings on all nodes:
```
[[RNode]]
type = RNodeInterface
enabled = True
outgoing = True
mode = gateway
port = /dev/ttyACM0
frequency = 867200000
bandwidth = 125000
txpower = 7
spreadingfactor = 8
codingrate = 5
```

With one client the transfer of pages works, as soon as the 2nd client requests a page at the same time it does not work anymore and gives the following 3 different errors. After these errors no single page request works anymore. After a restart of the client nomadnet it works again.

Test 1:
```
[2023-06-13 21:13:30] [Debug] Destination <d5a501ce3fe408858697f697038bafa7> is now 1 hops away via <d5a501ce3fe408858697f697038bafa7> on RNodeInterface[RNode]
[2023-06-13 21:13:31] [Verbose] Link establishment timed out
```

Test 2:
```
[2023-06-13 21:19:11] [Debug] Destination <d04f097e8845ba794747bdf8a7d063de> is now 1 hops away via <d04f097e8845ba794747bdf8a7d063de> on RNodeInterface[RNode]
[2023-06-13 21:19:12] [Extra] Registering link <fbea1b139a9e55191c7ad3ac6a08ef69>
[2023-06-13 21:19:12] [Debug] Link request <fbea1b139a9e55191c7ad3ac6a08ef69> sent to <nomadnetwork.node.669accff86278d771c105e59439c6e11/d04f097e8845ba794747bdf8a7d063de>
[2023-06-13 21:19:12] [Extra] Activating link <fbea1b139a9e55191c7ad3ac6a08ef69>
[2023-06-13 21:19:12] [Verbose] Link <fbea1b139a9e55191c7ad3ac6a08ef69> established with <nomadnetwork.node.669accff86278d771c105e59439c6e11/d04f097e8845ba794747bdf8a7d063de>, RTT is 0.875s
[2023-06-13 21:19:15] [Debug] Accepting resource advertisement for <4605b89f573f890e67142b0e700e4e6bf38ee1aef999c37ddfb13d616b19726a>
[2023-06-13 21:19:22] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:19:29] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:19:32] [Debug] Incoming response resource failed with status: 07
```

Test 3:
```
[2023-06-13 21:22:54] [Debug] Destination <d04f097e8845ba794747bdf8a7d063de> is now 1 hops away via <d04f097e8845ba794747bdf8a7d063de> on RNodeInterface[RNode]
[2023-06-13 21:22:54] [Extra] Registering link <ba547d37b6a9ec626781f0ccda8c7b48>
[2023-06-13 21:22:54] [Debug] Link request <ba547d37b6a9ec626781f0ccda8c7b48> sent to <nomadnetwork.node.669accff86278d771c105e59439c6e11/d04f097e8845ba794747bdf8a7d063de>
[2023-06-13 21:22:55] [Extra] Activating link <ba547d37b6a9ec626781f0ccda8c7b48>
[2023-06-13 21:22:55] [Verbose] Link <ba547d37b6a9ec626781f0ccda8c7b48> established with <nomadnetwork.node.669accff86278d771c105e59439c6e11/d04f097e8845ba794747bdf8a7d063de>, RTT is 0.776s
[2023-06-13 21:22:56] [Debug] Accepting resource advertisement for <7687dc2aabbf00b2167e781e0a36134771f018372a53eb6ede2bc19cc70f1271>
[2023-06-13 21:23:04] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:07] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:10] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:13] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:17] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:22] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:26] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:32] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:23:38] [Debug] Incoming response resource failed with status: 07
```

**Reply by SebastianObi** [reply]
Test 4:
Client 1 request page, transfer starts and progress is displayed. Then during this transfer client 2 request a page.
Transfer on client 1 freezes sometimes at 50% sometimes 75% and give the following error. Transfer on client 2 also failed.
Tranbsfer information with one client: 56,56KB 1,50KB in 5.28s 2.28/Kb/s

```
[2023-06-13 21:35:54] [Debug] Received page d04f097e8845ba794747bdf8a7d063de:/page/index.mu, caching for 12.000 hours.
[2023-06-13 21:35:54] [Debug] Cached page d04f097e8845ba794747bdf8a7d063de:/page/index.mu to /root/.nomadnetwork/storage/cache/a4a6bd6d867e6d56097f1c43d68b64530a004e0097946adda8df764fd3120650_1686731754.956873
[2023-06-13 21:36:05] [Extra] Valid announce for <8725888cf4d3fb9ae68f2cdb30dab685> 3 hops away, received via <a5622a99ff1b2e29f25403f5668fe154> on RNodeInterface[RNode] [RSSI -51dBm, SNR 12.0dB]
[2023-06-13 21:36:05] [Debug] Destination <8725888cf4d3fb9ae68f2cdb30dab685> is now 3 hops away via <a5622a99ff1b2e29f25403f5668fe154> on RNodeInterface[RNode]
[2023-06-13 21:36:06] [Debug] Removed /root/.nomadnetwork/storage/cache/a4a6bd6d867e6d56097f1c43d68b64530a004e0097946adda8df764fd3120650_1686731754.956873 from cache.
[2023-06-13 21:36:07] [Debug] Accepting resource advertisement for <6f2d39387618b0d595037bac58107e772c54cc80a4a632f0512c4679a9f3e3d1>
[2023-06-13 21:36:15] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:17] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:20] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:24] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:27] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:29] [Extra] Valid announce for <443310bf347df72769cbfd96b53fa51f> 3 hops away, received via <a5622a99ff1b2e29f25403f5668fe154> on RNodeInterface[RNode] [RSSI -51dBm, SNR 12.0dB]
[2023-06-13 21:36:29] [Debug] Destination <443310bf347df72769cbfd96b53fa51f> is now 3 hops away via <a5622a99ff1b2e29f25403f5668fe154> on RNodeInterface[RNode]
[2023-06-13 21:36:32] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:37] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:42] [Debug] Timed out waiting for parts, requesting retry
[2023-06-13 21:36:48] [Debug] Incoming response resource failed with status: 07
```

**Comment by SebastianObi** [comment]
Hi Mark,

I have tried to understand the rnode code. As you said CSMA is done. Is it possible that the waiting time after an rx is identical for all rnodes and then they start to retransmit at the same time?

I believe `lora_rx_turnaround_ms` is the waiting time, right?

At the following code:
```
    if (queue_height > 0) {
      if (!dcd_waiting) updateModemStatus();

      if (!dcd && !dcd_led) {
        if (dcd_waiting) delay(lora_rx_turnaround_ms);

        updateModemStatus();

        if (!dcd) {
          dcd_waiting = false;

          flushQueue();
          
        }
      } else {
        dcd_waiting = true;
      }
    }
```

Shouldn't it be a random value instead of a fixed value of 50ms?

**Comment by SebastianObi** [comment]
Has anyone tested reticulum with more than 2 rnodes in radio range?

It can't be only me who has these problems. Or is there something I am doing wrong?

A mesh protocol should be able to handle more than 2 nodes. Unfortunately in my scenario reticulum is completely unusable. :(

**Reply by faragher** [reply]
Just set up four RNodes, two sending telemetry every 15 seconds, one browsing NomadNet, and all bridged to the testnet, so consistently receiving announces. Everything is functioning here. The AOSP message board took 9 seconds to download at 2 Kb/sec, which is expected, and the telemetry radios were blocked during this, then properly broadcast their data when the channel was clear. I am not seeing any issues with the network. 

Two NomadNet page requests on two RNodes initially appears to function as intended, but with enough simultaneous requests and jammed frequencies, it appears that NomadNet stops serving pages over those specific links. This may be an issue with connections timing out under heavy load. Given time, it resets. I believe this may be due to the link somehow going stale or otherwise breaking. The radios appear to function flawlessly so long as the channel isn't busy for an excessive length of time. 

It *does* appear to be repeatable, though. If you request the same (large) resource from two requesters at once, one or both may fail. Every other asset on that server functions as intended, but that one requires a time-out before it functions again.

I think this is a problem with NomadNet serving, not Reticulum as a protocol. In either case, it does seem to need to be investigated. It's worth noting that if this is a time-out issue, it's related to duty cycle and transmit times. Investigating ways to space packets out more and if timeout delays are too short may be helpful.

**Reply by SebastianObi** [reply]
Sure, it needs further anaylsis.
Nomadnet uses the reticulum apis for the resource transfer. So it should be the reticulum links and resource transfer. It seems to be some timeouts issues.
Some time ago Mark said that the rnode does not use duty-cycle. So this should not be the problem.
The first thing to check is whether csma in the rnode is working without errors. This is the basic requirement for error-free communication on a shared medium. Then the behavior of the links/resources over slow half-duplex and shared media must be checked.

This scenario is certainly the worst case for a data transmission. This is a big challenge for reticulum.

**Reply by faragher** [reply]
I think you misunderstand, the problem isn't duty cycle, the solution is duty cycle. Since LoRa isn't multiple-access the way CDMA is multiple access, only one radio can be broadcasting at once. Broadcasting a large number of packets sequentially prevents anyone else from transmitting.  Without a duty cycle, a single radio can monopolize the frequency and thus nobody could transmit but them. Since the radios do, in fact, wait their turn to transmit, CSMA is working as intended. Also, in many areas, violating a certain duty cycle is illegal.

The time-out is a compromise from the word "go." As Mark has noted elsewhere, the data rate floor is due, in part, to time-outs. A slow bitrate requires a longer window than high speed transfer, and if you're running gigabit data rates, a multi-second time-out per bit creates other problems.

Fundamentally, the two things to investigate here are duty cycles and time-out parameters. Both are known issues to investigate, but the latter may be better dealt with in software. It's possible the separation of bit-rate related issues should happen higher than originally suggested, meaning RNode communications would have to know they required a slower time-out and thus stay open for longer.

Perhaps a link having a "high latency" flag that would drastically increase time-outs? Proper defaulting would make it backwards compatible and would likely solve this issue along with lowering minimum data rates.

**Reply by SebastianObi** [reply]
I am not an expert in this field, there must definitely be a solution to this problem. the fact is, this error can be easily reproduced and affects more than just one user. Only Mark will be able to help heri. He knows his software in detail. Where the problem might be and what a solution might look like.

In my test with the page transfer with a rnode sevrer and a rnode client my test page takes exactly 5 seconds. This can also be retransferred directly one after the other. In the best case then with 2 parallel clients it would take a little more than 10 seconds. But then it does not work at all. It does not seem to be retransmitted the data again. At some point there is no more activity on the rnodes. Although Nomandet is still waiting for a response.

I will open an issue for this becaus it's not a question.

**Reply by faragher** [reply]
Yes, there are solutions. I have proposed a few. Unfortunately, they're not user level. This is the problem with beta software. If you wish to continue with Mark, I have no issue with that, I was attempting to get you a faster solution and him the data needed to investigate.

The issue with the timing is exactly how long it takes. The nodes will dominate the frequency for the entire transmission, meaning one will most likely time out. However, if the second node does get a toe in, then it will attempt to broadcast all in one block, and then we end up stepping on the other radio. In short, as I said, even if nothing else is wrong, it's simply delaying the data too much and the system things something broke. 

Basically, they're not parallel clients, they're serial clients, but both expect their transmission to be the priority.

However, if you'd prefer to work only with Mark, that's your decision. Thank you for bringing the issue up.

**Reply by SebastianObi** [reply]
I didn't say I only wanted to work with someone in particular. It would certainly be easiest for the main developer to find the appropriate solution path. Anyone who can contribute to a solution and help is of course welcome :)

I only use the reticulum applications and apis for own applications. How this works internally in detail I am not the expert.

And of course i know that reticulm is beta software. It might have bugs and there are still optimizations to be done.
When i notice something by myself or have a suggestion for improvement i will be happy to share it here.

**Reply by faragher** [reply]
My apologies. I misinterpreted your intent.

NomadNet transfers pages over links. Looking at the link code, there are several timeouts, may of which are around 4-6 seconds. Now, I can't promise that the problem is those timeouts, the RNode only has so much buffer space, but five seconds over RNode isn't exactly a huge data transfer. While changing NomadNet or the link class is complicated, if you're simply using this as a basis for your own code, you can try some custom solutions.

I made a program which is basically Ping With Extra Steps. I implemented some duty cycle settings in the sender, and it functions well in a congested environment of similar systems, including some NomadNet nodes. 

https://github.com/faragher/RNode_Setting_Test/blob/main/Sender.py

There are some other ways to do this, as you can calculate the broadcast time in the API, but this is a pretty easy to understand way of doing this. It will only help if you're sending small bursts of data, but for custom applications, as a stop gap before there's a solution to the overarching issue. Sending data, packet by packet, over a link is more complicated than the base Reticulum utilities, but it allows more resilience. 

Now, base duty cycle in my test is 25%. Some countries have break points at 1% or 10%, but if it doesn't matter, 25%-50% is a good level, in my opinion, but it will decrease your data rate by the same factor. If you decide to use something like 90%, it will have a small increase in overhead, but it will allow other applications to send a packet in that gap. It might solve your issue.

Again, it's sadly only good for small transmission lengths, but if you use it packet by packet, or only for short messages, but for a custom application, it can prevent these kinds of jammed channels.

**Reply by SebastianObi** [reply]
Thanks for the quick reply and further advice.

I just use the existing APIs like in the examples to transfer files. Identical to NomadNet. The file is passed to Reticulum as a resource and the further processing/transfer is left to Reticulum. Of course it would be a possibility to develop an own transfer system with delays for the duty cycle etc.. That would be only an unclean workaround. It should be better tried to improve the existing resource transfer. But this will be more complicated for someone who does not know reticulum in detail.
